Sequence Component

christiemolloy opened the pull request:
This closes issue #767 

The sequence component will be used in the wizard and other components. It has two variations as per the Visual Design specs: a simpler one with three steps and only titles, and a more complex one with ten steps that can be navigated using the arrow controls, and title and status content. 

The sequence acts similar to the nav/tabs components. It is an unordered list of items and within each list item is a link that holds the position number and content. 

The Wizard Visual Design:
https://drive.google.com/drive/u/0/folders/1kegLVey0l7-8ym2K8kppYzIlnomXYLjl

patternfly-build commented:
Deploy preview for *pf-next* ready!

Built with commit ac978ec8f38412bae7a6a4111c2e9e4c400b2c05

https://deploy-preview-797--pf-next.netlify.com

andresgalante commented:
Hi @christiemolloy From the preview, and without reviewing the code I can see that the workspace is broken and there is a jump when I hover over the items.

It also only displays a dark background, this component should be color agnostic IMO

christiemolloy commented:
@andresgalante I turned off the default scroll-bar styling, could you let me know if that fixes the workspace? 

Working on fixing that jump on hover now, and also changing the colors on a white background, thanks!

jameslau commented:
@christiemolloy @andresgalante, scroll-bar styling is off now.

michael-coker commented on src/patternfly/components/Sequence/docs/code.md:
```diff
@@ -0,0 +1,32 @@
+| `.pf-c-sequence__content-status` | `<p>` | Initiates a sequence content status |
+| `.pf-m-hover` | `.pf-c-sequence__link` | Modifies to display the link as hovered |
+| `.pf-m-active` | `.pf-c-sequence__link` | Modifies to display the link as active |
+| `.pf-m-focus` | `.pf-c-sequence__link` | Modifies to display the link as focussed |
```
"focussed" should be "focused" :)

michael-coker commented on src/patternfly/components/Sequence/docs/code.md:
```diff
@@ -0,0 +1,32 @@
+| `.pf-c-sequence__list` | `<ul>` | Initiates a default sequence list |
+| `.pf-c-sequence__item` | `<li>` | Initiates a default sequence list item |
+| `.pf-c-sequence__link` | `<a>` | Initiates a default sequence list link |
+| `.pf-c-sequence__number` | `<div>` | Initiates a sequence number element |
```
looks like `.pf-c-sequence__number-container` needs to be added here.

christiemolloy replied:
added!

michael-coker commented on src/patternfly/components/Sequence/examples/sequence-complex-example.hbs:
```diff
@@ -0,0 +1,122 @@
+		{{/sequence-item}}
+		{{#> sequence-item sequence-item--attribute='aria-setsize="10" aria-posinset="7"'}}
+			{{#> sequence-link sequence-link--attribute='href="#"'}}
+				{{#> sequence-number-container}}
```
Just a general comment about `sequence-number-container`. The current/hover states cause the link width to change because the number size changes. IMO, the links shouldn't shift when you hover over or make a link active.

Also, if the sequence number will always be a number, I wonder if we could use a CSS counter and display the number as pseudo elements instead of in the markup. Just an idea.

christiemolloy replied:
I made a change to reduce the width of the spacer between the number container and content to be the width of the increase in width from the small number div to the large number div. So now it doesnt change on hover. Is this the right approach?

jgiardino replied:
> Also, if the sequence number will always be a number, I wonder if we could use a CSS counter and display the number as pseudo elements instead of in the markup. Just an idea.

@michael-coker Is that similar to the examples shown here: https://css-tricks.com/custom-list-number-styling/

Yesterday I had tested these examples in JAWS, plus [variations on lists](https://codepen.io/jenngiardino/pen/jeBreP?editors=1000.) in general. Some general observations:

- `<ol>` list items only announce the list item number if the `list-style` property isn't set to none.
- Navigating the css-tricks examples using JAWS resulted in separate stops for the counter and the list item contents. 

My recommendation would be to aim for having the list item counter and contents announced together, e.g. "link 1 Information". Doing this seems to be dependent on using inline elements, like `<a>` or `<span>`. But using `<div>` resulted in multiple keypresses, so we should change those to `<span>` elements if possible.

michael-coker replied:
@jgiardino yep that's it. thanks for testing that! good to know!

andresgalante commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,383 @@
+}
+
+// Override button styles
+button {
```
This is a global change, why is it here?

christiemolloy replied:
Thanks for picking that up. Is it correct to move it to _base.scss?

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,374 @@
+  --pf-c-sequence__link--PaddingRight: var(--pf-global--spacer--xl);
+  --pf-c-sequence__link--PaddingBottom: var(--pf-global--spacer--lg);
+  --pf-c-sequence__link--PaddingLeft: var(--pf-global--spacer--xl);
+  --pf-c-sequence__link--BorderBottomColor: var(--pf-global--Color--200);
```
Per the design, this should be semi transparent.

christiemolloy replied:
Turns out, when I change the colors to work for both light/dark themes that these colors actually become white, to pass accessibility tests, so im not sure if transparency will be in the picture

jgiardino commented on src/patternfly/components/Sequence/examples/sequence-finite-steps.hbs:
```diff
@@ -0,0 +1,37 @@
+{{#> sequence sequence--modifier='aria-label="Sequence"'}}
+	{{#> sequence-list sequence-list--attribute='type="1"'}}
```
[MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ul#Attributes) lists the `type` attribute as deprecated.

christiemolloy replied:
thanks that was left there from when i was testing <ol>

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,374 @@
+  --pf-c-sequence__content--title--MaxWidth: 100px;
+
+  // Content Title Color
+  --pf-c-sequence__content-title--Color: var(--pf-global--Color--light-100);
```
Per the design, shouldn't this match the circle/bottom-border color? I also can't tell for certain in the PDF, but it looks like the text and circle are semi-transparent, too (like the bottom border)

christiemolloy replied:
In the sketch file it looks like the title color is white, and the status color is different to the number color. The number color is a solid, but you are right the border color is transparent so will change that

jgiardino commented on src/patternfly/components/Sequence/docs/code.md:
```diff
@@ -0,0 +1,33 @@
+| -- | -- | -- |
+| `aria-label="[landmark description]"` | `.pf-c-sequence` | Describes <sequence> landmark |
+| `aria-setsize` | `.pf-c-sequence__item` | Indicates the total number of items in the list |
+| `aria-posinset` | `.pf-c-sequence__item` | Indicates the position of an item within a list |
```
I was hoping that `aria-setsize` and `aria-posinset` would provide some indication about which list item currently has focus, but when I tested this in JAWS and VO, nothing was announced. Did you notice any change when you added this in? At this point, I'm leaning toward not including this, until we find that there is support and it provides some benefit to users.

christiemolloy replied:
In VO nothing was announced when I tested it either. I will leave it off for now

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,374 @@
+  // Item
+  --pf-c-sequence__item--MarginRight: 1px;
+  --pf-c-sequence__item--MarginLeft: 1px;
+  --pf-c-sequence__item--MaxWidth: 200px;
```
Can you change this to `#{pf-size-prem(200px)};`?

christiemolloy replied:
yes, changed this

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,374 @@
+.pf-c-sequence__item {
+  max-width: var(--pf-c-sequence__item--MaxWidth);
+  margin-right: var(--pf-c-sequence__item--MarginRight);
+  margin-left: var(--pf-c-sequence__item--MarginLeft);
```
just curious why you use left/right margin, and not something like `&:not(:last-child) { margin-right: var(--pf-c-sequence__item--MarginRight); } ` and change the var to 2px.

christiemolloy replied:
added &:not(:last-child) for margin-right and added &:not(:first-child) for margin-left, to ensure there are no extra margins. IMO changing the var to 2px creates too large of a space in between them as compared to the docs. What was your thought behind changing it to 2px?

michael-coker replied:
oh sorry... what I mean is you're adding a margin-left and margin-right on all of the elements, which creates a 2px gap between them. why not just use a 2px margin-right on all but the last element, or 2px margin-left on all but the first?

christiemolloy replied:
ohh thanks for clarifying! making those changes

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,374 @@
+// Link
+.pf-c-sequence__link {
+  display: flex;
+  justify-content: flex-start;
```
the default is `flex-start` so I don't think this is necessary

christiemolloy replied:
good find, thanks

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,374 @@
+  --pf-c-sequence__control--BorderBottomHeight: var(--pf-global--BorderWidth--lg);
+  --pf-c-sequence__control--BorderTopColor: transparent;
+  --pf-c-sequence__control--BorderBottomColor: transparent;
+  --pf-c-sequence__control--hover--BorderTopHeight: var(--pf-global--BorderWidth--lg);
```
technically speaking, the `Border...Height` vars should be `Border...Width` since `border-width` is the property you're modifying.

christiemolloy replied:
replaced all Border...Height's with Border...Width

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,374 @@
+  --pf-c-sequence__content--MinHeight: var(--pf-global--spacer--xl);
+
+  // content Title
+  --pf-c-sequence__content--title--MaxWidth: 100px;
```
Can you change this to `#{pf-size-prem(100px)};`?

christiemolloy replied:
yes changed this

christiemolloy commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,374 @@
+  // added so that the list will be read by screen reader
+  &::before {
+    width: 1px;
+    content: "";
```
I did style it like above, however it was added extra spacing and shifting the design a lot. I was hoping it would work with any content but it looks like "\200B" is necessary

christiemolloy replied:
I think it would be a good discussion for the group, especially since it does affect styling and is an extra step to get working

christiemolloy replied:
I added it back in, and it reads: "list three items" but only after it reads the content of the first list item

jgiardino replied:
Weird. Thanks for testing this.

christiemolloy replied:
@jgiardino from the a11y discussion, Ill remove: 

ul li::before {
  content: "\200B";
}

and add in a a role for the list and for the list item?

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,374 @@
+  }
+}
+
+.pf-c-sequence__control-left {
```
instead of `left` and `right`, I would use a more [logical property-esque](https://www.w3.org/TR/css-logical-1/) name like `start` and `end`

michael-coker replied:
And I wonder if this should just be `.pf-c-sequence__control`, since they're the same element with different presentations, then use `.pf-m-start` and `.pf-m-end` as modifiers. Similar to the `.pf-c-switch__label` element in the Switch component.

jgiardino replied:
sounds good!

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,364 @@
+  // Number Border
+  --pf-c-sequence__number--BorderColor: var(--pf-global--Color--200);
+  --pf-c-sequence__number--BorderRadius: var(--pf-global--BorderRadius--lg);
+  --pf-c-sequence__number--BorderWeight: var(--pf-global--BorderWidth--md);
```
`BorderWeight` should be `BorderWidth`

christiemolloy replied:
thanks, updated

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,364 @@
+}
+
+// Vertically centers number and border
+.pf-c-sequence__number-container {
```
Is this element necessary? Just glancing at it, I think you can remove it and just have `.pf-c-sequence__number`, then add `align-items: center;` to `.pf-c-sequence__link` to vertically center it.

christiemolloy replied:
thanks that works :)

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,364 @@
+  color: var(--pf-c-sequence__number--Color);
+  text-align: center;
+
+  &::before {
```
Why is the border in a pseudo element? Would it work applied to `.pf-c-sequence__number`?

christiemolloy replied:
The main reason for the border being a pseudo element was that when the sizing changed, the border wouldnt interfere with the padding changes etc. but now that we are using transform and scale i believe we dont need this pseudo element, going to check if it works

christiemolloy replied:
it works, so removing this pseudo element and adding border to the number

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,364 @@
+
+// Content Title
+.pf-c-sequence__content-title {
+  display: block;
```
I don't think this is necessary since the parent is `flex`

christiemolloy replied:
removed :)

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,364 @@
+// Content Status
+.pf-c-sequence__content-status {
+  font-size: var(--pf-c-sequence__content-status--FontSize);
+  color: var(--pf-c-sequence__content-status--Color);
```
If you put a lot of text in this element, it pushes the parent link's border down.

christiemolloy replied:
Do you think a good option would be to truncate it? Since its a status, im assuming that means it shouldnt really be more than a few words

christiemolloy replied:
@michael-coker

michael-coker replied:
That seems fine to me, since you're truncating the title text, too. You can always ask one of the designers as well how they expect the element to behave with more/less text.

christiemolloy replied:
truncated for now, will double check with Kyle when he reviews

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,364 @@
+      // min-width: var(--pf-c-sequence__number--hover--MinWidth);
+      // max-height: var(--pf-c-sequence__number--hover--MaxHeight);
+      // margin-right: var(--pf-c-sequence__number--hover--MarginRight);
+      // font-size: var(--pf-c-sequence__number--hover--FontSize);
```
can you remove the commented out lines, and the rest below?

christiemolloy replied:
yes

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,364 @@
+
+  &.pf-m-hover,
+  &:hover {
+    border-bottom: var(--pf-c-sequence__link--BorderBottomHeight) solid var(--pf-c-sequence__link--hover--BorderBottomColor);
```
since you don't change the `border-width` in any of these states, can you rewrite it just to change the `border-color` property?

christiemolloy replied:
good suggestion, fixed

michael-coker commented on src/patternfly/_base.scss:
```diff
@@ -153,3 +163,12 @@ html {
 } // ends if
+
+// remove scrollbar space
+::-webkit-scrollbar {
```
`_base.scss` is the base stylesheet for PF4. When this stylesheet is added to a site, the styles in it will apply to everything on the site. For example, the workspace no longer has a scrollbar. All buttons (unless there is specific CSS for the button) will no longer have padding, bbackground, border, etc. I know you just moved these from the sequence stylesheet, but did you mean to limit the scope of these changes to the sequence component?

michael-coker replied:
Yes, but move them inside of the `.pf-c-sequence { // in here somewhere }` selector block. I believe before you had added them outside of that block, which will apply to the entire site (similar to how adding it to `_base.scss` applies to the whole site). And if the styles for `button` are for the sequence control element, apply the CSS to `.pf-c-sequence__control`, not `button`

christiemolloy replied:
thanks for clarifying that - moved them to their respective areas

michael-coker commented on src/patternfly/components/Sequence/docs/code.md:
```diff
@@ -0,0 +1,33 @@
+| `.pf-c-sequence__number-container` | `<div>` | Initiates a sequence number container element |
+| `.pf-c-sequence__number` | `<div>` | Initiates a sequence number element |
+| `.pf-c-sequence__content` | `<div>` | Initiates a sequence content element |
+| `.pf-c-sequence__content-title` | `<h1>, <h2>, <h3>, <h4>, <h5>, <h6>` | Initiates a sequence content title |
```
this is applied to a `span` in your examples.

christiemolloy replied:
changed

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,363 @@
+  --pf-c-sequence__control--PaddingRight: var(--pf-global--spacer--sm);
+  --pf-c-sequence__control--PaddingBottom: var(--pf-global--spacer--xl);
+  --pf-c-sequence__control--PaddingLeft: var(--pf-global--spacer--sm);
+  --pf-c-sequence__control--BorderTopHeight: var(--pf-global--BorderWidth--lg);
```
I might use a pseudo element for this border, so that you can just remove a border when you don't want to show it without causing the content in the border to shift. The edges where the borders meet looks weird now

<img width="120" alt="screen shot 2018-10-10 at 4 02 05 pm" src="https://user-images.githubusercontent.com/35148959/46765904-f3fe4e80-cca5-11e8-8727-3793ca206a04.png">

christiemolloy replied:
I added the border to a before pseudo element and that fixes that weird style where the borders meet.

jgiardino commented on src/patternfly/components/Sequence/docs/code.md:
```diff
@@ -0,0 +1,33 @@
+| `.pf-c-sequence__link` | `<a>` | Initiates a default sequence list link |
+| `.pf-c-sequence__number-container` | `<div>` | Initiates a sequence number container element |
+| `.pf-c-sequence__number` | `<div>` | Initiates a sequence number element |
+| `.pf-c-sequence__content` | `<div>` | Initiates a sequence content element |
```
Is it possible to change these elements to `<span>`? When I tested in JAWS, it seems to see `<div>` elements as separate chunks of content which results in more key presses for scanning the contents—one for the number and one for the label. Ideally, the contents of each list item / link would be announced as a result of a single keypress.

| Class | Applied To | Outcome |
| -- | -- | -- |
| `.pf-c-sequence__number-container` | `<div>` | Initiates a sequence number container element |
| `.pf-c-sequence__number` | `<div>` | Initiates a sequence number element |
| `.pf-c-sequence__content` | `<div>` | Initiates a sequence content element |

jgiardino commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,363 @@
+  text-align: center;
+
+  &::before {
+    position: absolute;
```
This is a great question! We should discuss this at our a11y meeting since this probably mostly affects screen reader users and keyboard-only users.

michael-coker replied:
@jgiardino sounds good! and we can always leave the source order as is if needed, still use non-absolute positioning, and use flex `order` to re-arrange visually.

christiemolloy replied:
Taken absolute position off of the controls and they are placed within the sequence using flex order to re-arrange them visually.

jgiardino replied:
I think this [codepen](https://codepen.io/benfrain/full/zZZLaP/) was the source of inspiration for this design. When I tested this example using only a keyboard and also using voiceover, the arrow buttons disappear before I would be able to place focus on them, because in both cases, the tabs scroll to the end as I shift focus. 

Because of this, I don't think DOM order is that important, since they're not needed for keyboard-only navigation. 

But we still want to keep the `aria-label` attributes as you have them for screen reader users who are using a mouse or touch.

michael-coker replied:
Also can we set a min-width on these buttons so that the target area abides by the 44x44px minimum? the var for that is `--pf-global--target-size--MinWidth`

christiemolloy replied:
yep, just updated that

christiemolloy commented on src/patternfly/components/Sequence/examples/sequence-complex-example.hbs:
```diff
@@ -0,0 +1,122 @@
+{{#> sequence sequence--modifier='aria-label="sequence"'}}
```
good catch, thanks

matthewcarleton commented on src/patternfly/components/Sequence/docs/code.md:
```diff
@@ -0,0 +1,32 @@
+## Overview
+
+Only if needed write a short description with implementation notes. Design and interaction notes are already written on the design specs, don't repeat information.
```
Can you fill in a brief summary here

christiemolloy replied:
sure!

matthewcarleton commented on src/patternfly/components/Sequence/examples/sequence-example.hbs:
```diff
@@ -0,0 +1,102 @@
+  {{#> sequence-control sequence-control--modifier="pf-c-sequence__control-m-start"}}
+		{{#> sequence-control-icon sequence-control-icon--start="true" sequence-control-icon--attribute='aria-label="Start"'}}
+		{{/sequence-control-icon}}
+	{{/sequence-control}}
```
can you fix the indents for this file.

christiemolloy replied:
yep!

matthewcarleton commented on src/patternfly/components/Sequence/sequence-content.hbs:
```diff
@@ -0,0 +1,6 @@
+<span class="pf-c-sequence__content{{#if sequence-content--modifier}} {{sequence-content--modifier}}{{/if}}"
```
nitpicky but wondering why this is a span? It seems more like a div to me if it's grouping things.

christiemolloy replied:
Changed to a span so that the screen reader would read the content in order rather than breaking it up into chunks from the divs

matthewcarleton replied:
ah ok!

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,344 @@
+  --pf-c-sequence__control--BorderBottomWidth: var(--pf-global--BorderWidth--lg);
+  --pf-c-sequence__control--BorderTopColor: transparent;
+  --pf-c-sequence__control--BorderBottomColor: transparent;
+  --pf-c-sequence__control--hover--BorderTopWidth: var(--pf-global--BorderWidth--lg);
```
these should be organized by their state not their property, I think it makes it easier to scan that way.

christiemolloy replied:
just rearranged that, let me know if theres anything else I should rearrange to make it easier to scan

matthewcarleton replied:
👍

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,344 @@
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  min-width: var(--pf-c-sequence__number--MinWidth);
```
I've noticed that with a number with two digits the circle grows. In the design it shows "10" as still in a circle. With this solution the circle will become an oval. This could be brought back to design if we don't have a good solution. I know we investigated the psuedo element concept a bit but I'm not sure where that landed.

christiemolloy replied:
I did see the weird styling with the 10. What Im thinking right now is if we change min-width and max-height to width and height, then there is no room for the circle to change to an oval shape, when the content gets too wide. What do you think @matthewcarleton

matthewcarleton replied:
ya this is a tricky one to solve, if the font size grows then it breaks. Which isn't good for accessibility.

christiemolloy replied:
@matthewcarleton is it possible to make the width and height % of the font size.

christiemolloy commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,344 @@
+    border-bottom: var(--pf-c-sequence__control--focus--BorderBottomWidth) solid var(--pf-c-sequence__control--focus--BorderBottomColor);
+  }
+
+  button {
```
I think this is removed now

christiemolloy commented:
@mcarrano thanks for your comment. I've combined the two variations into one. I changed the component to hide the control buttons at a smaller viewport size. However I have left both of the control buttons on, because I dont think its possible in html/css to hide/show a button based on the position of the scroll or the number of items. However if we want to include javascript in here to demonstrate how the sequence should really function, then that could be a conversation. Also working to remove the scrollbar styling in mozilla, its a little trickier!

mcarrano commented:
@christiemolloy Thanks for making this change. Yes, it's probably not possible to fully demonstrate this in the HTML/CSS version. But one thing - I think you have the logic for hiding the control buttons is reversed. As a shrink the viewport size, I expect the buttons to show, not hide. At larger sizes, I don't need the scroll to see the entire sequence bar, right? It might help to make the example shorter (less steps) so the entire bar will fit at full width. Then one can see the buttons revealed at the point where a scroll bar would be needed. Does that make sense?

christiemolloy commented:
@mcarrano sorry I was confused with the visual designs because it looks like at a large viewport the buttons appear and at a small viewport they disappear - like on page 9? But your recommendation makes sense-- going to lesson the number of steps and have the buttons only appear when the viewport shrinks.

jgiardino commented:
Here are some notes based on reviewing various list examples with screen readers:

1. In VO, `role="list"` helps, but `role="listitem"` made no difference. Let's keep `role="list"` but remove `role="listitem"` to simplify our markup.
2. I tested this [codepen](https://codepen.io/benfrain/full/zZZLaP/) in JAWS, which is similar to the design we're using. I noticed that the tabs will scroll to the left as JAWS focus moves through the tabs. Unfortunately, the tabs will not scroll to the right as JAWS focus moves through the tabs to the left. And the arrow buttons are visible to the screen reader, but not labeled and not in a DOM order that matches layout. They’re positioned at the end of the tabs, so leaving the last tab, I hear “unlabeled one button” and “unlabeled two button”. They’re not labeled for the screen reader or hidden from the screen reader. 
    1. My suggestion would be to place the buttons in the DOM to match the order they appear in the layout (e.g. start button before the list items and end button after the list items). This way, if the screen reader user does need them, they're present in a logical order.
3. Previously, mcoker suggested using a method similar to what’s documented here: https://css-tricks.com/style-list-markers-css/ When I originally tested this, the number and contents were announced separately. But when I forked the [codepen](https://codepen.io/jenngiardino/pen/XxaqBN) and changed the marker style from inline-block to inline, it was announced together. This means that:
    1. This method is a valid option for this component if you think it's beneficial
    2. `display:inline-block` vs `display:inline` will affect how elements are announced (e.g. with separate keypresses or together), at least in JAWS.

christiemolloy commented:
@jgiardino thanks for testing this. 

1. I removed role="listitem"
2. I changed the aria-label from the icon to the button - I'm thinking that this may have an impact on how the button was read out. If you think so too, it might be worth checking again. I did also move them in order of how they are positioned on the page. I did give them a flex order though but I guess that doesn't work.
3. That is good to know. I think this style will be difficult to implement and will take time. So I'm going to stick with what we have now. Is it reading better now after spans replaced with divs.

jgiardino commented:
Thanks for the updates, @christiemolloy!

> Is it reading better now after spans replaced with divs.

JAWS seems to be affected by weird white space and line breaks. I noticed when I cleaned up the html that I pasted into a codepen, that it would announce the `<li>` children as a single list item.

NVDA on the other hand does not. I don't think that should block this PR from moving forward. The only solution I can think of is to apply an `aria-labelledby` on the `<a>` and have it reference itself or the individual text elements inside it to see if that results in the list item being announced with a single keypress. But I'd prefer to test that and get input from others on this before suggesting any changes.

christiemolloy commented:
@michael-coker thankyouuuu!

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,350 @@
+  }
+}
+
+.pf-c-sequence__control {
```
I'm not sure we need lines 143-146. The padding is removed above on line 120 and top and bottom aren't doing anything. Line 120, doesn't seem to doing anything either unless this is browser specific.

christiemolloy replied:
yeah those aren't affected by different browsers so I can remove

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,350 @@
+
+// Content Title
+.pf-c-sequence__content-title {
+  max-width: var(--pf-c-sequence__content-title--MaxWidth);
```
I've noticed that if I add a longer string of text to the title the number is no longer keeping it's size. I think this can be fixed by adjusting the max-width. Let me know if you want to chat about a fix for this.

christiemolloy replied:
Increased the max-width on the title/status and the item to allow for larger strings of text. For now this follows the design, however if anyone feels strongly about having the width of the items be the same, then we can definitely revise.

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,350 @@
+}
+
+// Content
+.pf-c-sequence__content {
```
you can save some css lines if you add this here instead of each of the children:
```
& > * {
    max-width: var(--pf-c-sequence__content--MaxWidth);
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
```

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,350 @@
+
+  position: relative;
+  display: flex;
+  justify-content: center;
```
Is this needed?

christiemolloy replied:
no, removed

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,350 @@
+
+.pf-c-sequence__control-icon {
+  min-width: var(--pf-c-sequence__control-icon--MinWidth);
+  color: var(--pf-c-sequence__control-icon--Color);
```
I think you get this for free so it's not needed here.

christiemolloy replied:
yep!

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,339 @@
+
+.pf-c-sequence__control-m-start {
+  left: var(--pf-c-sequence__control-m-start--PositionLeft);
+  order: 1;
```
remove?

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,339 @@
+}
+
+.pf-c-sequence__control-m-end {
+  right: var(--pf-c-sequence__control-m-end--PositionRight);
```
remove this and the order: 3;?

christiemolloy replied:
yep

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,339 @@
+// Horizontal list
+.pf-c-sequence__list {
+  display: inline-flex;
+  order: 2;
```
Remove?

christiemolloy replied:
yep

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,339 @@
+  border-bottom: var(--pf-c-sequence__link--BorderBottomWidth) solid var(--pf-c-sequence__link--BorderBottomColor);
+
+  &::before {
+    position: relative;
```
Do you need this?

christiemolloy replied:
nope :)

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,339 @@
+    position: relative;
+    width: var(--pf-c-sequence__link--Width);
+    height: var(--pf-c-sequence__link--Height);
+    margin-right: var(--pf-c-sequence__link--MarginRight);
```
Do you see the odd overlap that this margin creates when you inspect it? I don't know if it's hurting anything 🤔

christiemolloy replied:
I do see it when the number is on hover, active, focus or current. Margin-right didnt need to be changed when the number gets bigger, because with transform it doesnt affect the other positioning. I could change the spacer size, but I wonder if that creates more room for error when the text size changes and the text might jitter. Thoughts? @matthewcarleton

matthewcarleton commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,339 @@
+  --pf-c-sequence__control--focus--BorderBottomColor: var(--pf-global--primary-color--100);
+
+  // Sequence Control Modifier Start
+  --pf-c-sequence__control-m-start--PositionLeft: 0;
```
This can be removed as well as a couple others once you update the changes below.

christiemolloy replied:
yep removed

christiemolloy commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,348 @@
+  --pf-c-sequence__content-status--FontSize: var(--pf-global--FontSize--xs);
+  --pf-c-sequence__content-status--FontWeight: var(--pf-global--FontWeight--regular);
+
+  position: relative;
```
yeah its not needed

christiemolloy commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,348 @@
+}
+
+.pf-c-sequence__control-m-start {
+  left: var(--pf-c-sequence__control-m-start--PositionLeft);
```
yep, removed

christiemolloy commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,348 @@
+  display: flex;
+  justify-content: center;
+  height: var(--pf-c-sequence--Height);
+  overflow-y: hidden;
```
Oh wow I didn't even notice it was cut off. Added 1px to the height and now it works.

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,309 @@
+  overflow-y: hidden;
+
+  // remove default border styling
+  button {
```
I think all of the CSS here should be under `.pf-c-sequence__control`?

christiemolloy replied:
When I move it to sequence__control those button styles dont show, so moved it back to the sequence

michael-coker replied:
Hmm, I tested this on my end and it seems to work, but maybe I'm missing something? I created a PR if you want to check it out - https://github.com/christiemolloy/patternfly-next/pull/4

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,309 @@
+  }
+}
+
+.pf-c-sequence__control-m-start {
```
this should be `.pf-m-start` (and `.pf-m-end` below), and nested inside of `.pf-c-sequence__control`

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,309 @@
+
+// Change default styles for the list and list items
+.pf-c-sequence__list .pf-c-sequence__item {
+  list-style-type: none;
```
can this just be listed under `.pf-c-sequence__item`?

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,309 @@
+
+  display: flex;
+  justify-content: center;
+  height: var(--pf-c-sequence--Height);
```
Why is there a height?

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,309 @@
+  button {
+    background-color: transparent;
+    border-top: 0;
+    border-bottom: 0;
```
I think you can replace these borders with `border: 0;`, then just add a `border-right` on the start button and `border-left` on the end button

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,309 @@
+// Horizontal list
+.pf-c-sequence__list {
+  display: inline-flex;
+  height: var(--pf-c-sequence__list--Height);
```
is this height necessary?

christiemolloy replied:
The height on here and the sequence <nav> is so that the scroll could be hidden in firefox

michael-coker commented on src/patternfly/components/Sequence/sequence.scss:
```diff
@@ -0,0 +1,309 @@
+
+  // Content Status Font
+  --pf-c-sequence__content-status--FontSize: var(--pf-global--FontSize--xs);
+  --pf-c-sequence__content-status--FontWeight: var(--pf-global--FontWeight--regular);
```
is this necessary?

michael-coker commented:
Looking good! I left a few more comments, let me know if you have questions.

srambach commented:
If I add more steps, then the arrows to the left and right do not necessarily come in as soon as the steps overflow. They seem to be hardcoded to an xl breakpoint. Is that the intended behavior?

christiemolloy commented:
@srambach I have removed the hardcoded break-point that determined when the arrows would appear/disappear. React will handle when these arrows appear based on the number of items in the sequence.
